keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,344 @@
+
+### Risks and Mitigations
+
+
+
+Existing users may be impacted as we're enforcing to use RuntimeClasses instead of plain nodeSelector and tolerations.
+
+
+The downside of this approach is since we're making RuntimeClass as default choice for the users, existing users have to update their pod spec or workload pod template spec to ensure that their workloads are not broken during upgrades. 
+This can be mitigated by 
+- Having appropriate warnings in the releases before we beta and GA will mitigate the problem.
+- Having an example out-of-tree admission webhook which can detect such workloads and update them.(Having an in-tree admission plugin which mutates the pod spec to have appropriate Windows specific RuntimeClass may be too hard to do generically and not transparent to the user)
+
+
+## Design Details
+
+
+We can piggyback on the existing RuntimeClass admission controller to query for the RuntimeClass and see if it has
+`kubernetes.io/os: windows` in it's scheduling field. 
+TODO: We need to get an ack from sig-auth to use it.
+
+If we cannot piggyback on the RuntimeClass controller we need to build our own validating admission controller for Windows specific pods.
+
+### Test Plan
+
+
+Whatever path we choose, unit and e2e tests are a hard requirement to ensure smooth
+transition to make RuntimeClasses default choice for the users.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+- Users getting a warning to switch to RuntimeClasses in beta release
+
+#### Alpha -> Beta Graduation
+- Gather feedback from end users
+- Tests are in Testgrid and linked in the KEP
+- Deny admission to pods that have only `kubernetes.io/os: windows` in nodeSelector
+  but don't have an associated RuntimeClass with `kubernetes.io/os: windows` in
+  scheduling field.
+
+#### Alpha -> Beta Graduation
+- 2 examples of end users using this field
+
+### Upgrade / Downgrade Strategy
+
+- Upgrades:
+  When upgrading from a release without this feature, to a release with `RuntimeClassesForWindows` enabled, we will honor both nodeSelector and RuntimeClasses
+  but a warning will be sent to the user that `honoring explicit nodeSelector without Runtimeclasses will be stopped from next release` if the pod has just nodeSelector set. This ensures users have enough time to move without breaking their workloads.
+- Downgrades:
+  When downgrading from a release with this feature to a release without `RuntimeClassesForWindows`, the existing behavior will continue where both nodeSelector and RuntimeClasses are honoured without a warning message.
+
+### Version Skew Strategy
+
+If the feature gate is enabled, only admission controller(either RuntimeClass or Windows specific admission controller) should be impacted. This feature may have some kubelet implications as the code to strip security constraints based on OS can be removed.
```

Comment:
This feature will have kubelet implications

Reply:
Should we call out `api-server` here also as per @marosset's [comment](https://github.com/kubernetes/enhancements/pull/2803/commits/3a90d782341954e121941eb747f3918217806c5b#r669944486)?

Reply:
This enhancement was discussed at sig-node today.
One suggestion make by @derekwaynecarr was it might make more sense to strip the security constraints for that don't apply to the intended OS at admission time rather than have the kubelet remove the fields. That way we wouldn't be persisting unnecessary data and also could help backwards compatibility.

Reply:
i am strongly in favor of making it clear that a pod is authoritatively a windows or linux pod.
i am not sure i like a strong correlation to runtime class as the mechanism for end-users.
i still prefer the idea of a dedicated field on the pod, and regret we did not do what oci spec did by having a dedicated linux/windows section in the pod spec.  it makes it less cumbersome just looking at the pod resource what type of pod this pod is intending to produce.  i also would like to couple kubelet feature support specific to windows with a dedicated field for things that are not supported on windows (ephemeral containers, emptyDir, etc.) that could be supported potentially in the future.  right now, we just admit these pods but fail at runtime is my understanding.

can we add a specific section discussing the negative of an explicit field in the pod spec?